Sources/SwiftGRPCNIO/GRPCClientChannelHandler.swift

```diff
@@ -0,0 +1,200 @@
+  public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+    guard self.inboundState != .ignore else { return }
+
+    switch unwrapInboundIn(data) {
+    case .headers(let headers):
+      guard self.inboundState == .expectingHeaders else {
+        self.errorCaught(ctx: ctx, error: GRPCStatus.processingError)
+        return
+      }
+
+      self.initialMetadataPromise.succeed(result: headers)
+      self.inboundState = .expectingMessageOrStatus
+
+    case .message(let message):
+      guard self.inboundState == .expectingMessageOrStatus else {
+        self.errorCaught(ctx: ctx, error: GRPCStatus.processingError)
+        return
+      }
+
+      self.responseObserver.observe(message)
+
+    case .status(let status):
+      guard self.inboundState == .expectingMessageOrStatus else {
+        self.errorCaught(ctx: ctx, error: GRPCStatus.processingError)
+        return
+      }
+
+      self.observeStatus(status)
+
+      // We don't expect any more requests/responses beyond this point.
+      self.close(ctx: ctx, mode: .all, promise: nil)
+    }
+  }
+}
+
+extension GRPCClientChannelHandler: ChannelOutboundHandler {
+  public typealias OutboundIn = GRPCClientRequestPart<RequestMessage>
+  public typealias OutboundOut = GRPCClientRequestPart<RequestMessage>
+
+  public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+    guard self.inboundState != .ignore else { return }
+
+    switch unwrapOutboundIn(data) {
+    case .head:
+      guard self.outboundState == .expectingHead else {
+        self.errorCaught(ctx: ctx, error: GRPCStatus.processingError)
+        return
+      }
+
+      // See the documentation for `requestHeadSentPromise` for an explanation of this.
+      self.requestHeadSentPromise = promise ?? ctx.eventLoop.newPromise()
+      ctx.write(data, promise: self.requestHeadSentPromise)
+      self.outboundState = .expectingMessageOrEnd
+
+    default:
+      guard self.outboundState == .expectingMessageOrEnd else {
+        self.errorCaught(ctx: ctx, error: GRPCStatus.processingError)
+        return
+      }
+
+      ctx.write(data, promise: promise)
+    }
+  }
+}
+
+extension GRPCClientChannelHandler {
+  /// Closes the HTTP/2 stream. Inbound and outbound state are set to ignore.
+  public func close(ctx: ChannelHandlerContext, mode: CloseMode, promise: EventLoopPromise<Void>?) {
+    self.observeStatus(GRPCStatus.cancelled)
+
+    requestHeadSentPromise.futureResult.whenComplete {
+      ctx.close(mode: mode, promise: promise)
+    }
+
+    self.inboundState = .ignore
+    self.outboundState = .ignore
+  }
+
+  /// Observe an error from the pipeline. Errors are cast to `GRPCStatus` or `GRPCStatus.processingError`
+  /// if the cast failed and promises are fulfilled with the status. The channel is also closed.
+  public func errorCaught(ctx: ChannelHandlerContext, error: Error) {
+    let status = (error as? GRPCStatus) ?? .processingError
```

Comment:
Do we want to forward errors to a delegate, similar to your server-side PR? Feel free to move this to a follow-up PR, though, in that case would you mind adding a TODO?

Reply:
I planned on doing a follow-up PR. I'll add a note here.

Reply:
👍